### PR TITLE
Drivers: DMIC: Updates to enable runtime instantiate and remove DAI

### PR DESCRIPTION
**Pull request by singalsu**
This pull request replaces incorrectly done #3592 where the PR was created from a SOF branch and not my own forked repo.

I'll keep this as draft for now. @ranj063 I think it's up to you to decide if this change is needed.

### CODE REVIEW

File: src/drivers/intel/dmic.c [hunk 1]
```diff
@@ -579,9 +578,22 @@ static int select_mode(struct dai *dai,
 	else
 		mfir = modes->mfir_b;
 
-	mmin = find_min_int16(mfir, modes->num_of_modes);
-	count = find_equal_int16(idx, mfir, mmin, modes->num_of_modes, 0);
-	n = idx[count - 1];
+	/* Search fir_list[] decimation factors from start towards end. The found
+	 * last configuration entry with searched decimation factor will be used.
+	 */
+	count = 0;
```
Comment (lgirdwood) [inline src/drivers/intel/dmic.c:584]: this is really an array index we are looking for ? I would rename this ```found```

File: src/drivers/intel/dmic.c [hunk 2]
```diff
@@ -579,9 +578,22 @@ static int select_mode(struct dai *dai,
...
+					 modes->num_of_modes, 0);
+		if (count > 0)
+			break;
+	}
+
+	if (!count) {
+		dai_err(dai, "select_mode(): No filter for decimate by %d found", mfir);
+		return -EINVAL;
+	}
+	n = idx[count - 1]; /* Option with highest clock divisor and lowest mic clock rate */
```
Comment (lgirdwood) [inline src/drivers/intel/dmic.c:596]: and this is not needed since ```n``` should already have the correct value ?

**Review by lgirdwood — CHANGES_REQUESTED** [review]

File: tools/topology/development/sof-tgl-nocodec-ci.m4 [hunk 3]
```diff
@@ -242,7 +242,7 @@ DAI_CONFIG(DMIC, 0, 6, NoCodec-6,
 	   dnl DMIC_CONFIG(driver_version, clk_min, clk_mac, duty_min, duty_max,
 	   dnl		   sample_rate, fifo word length, unmute time, type,
 	   dnl		   dai_index, pdm controller config)
-	   DMIC_CONFIG(1, 500000, 4800000, 40, 60, 48000,
+	   DMIC_CONFIG(1, 2400000, 4800000, 40, 60, 48000,
```
Comment (ranj063) [inline tools/topology/development/sof-tgl-nocodec-ci.m4:245]: @singalsu can you please rebase this PR? I think this patch may no longer be needed after PIerre's changes

Comment (singalsu) [reply tools/topology/development/sof-tgl-nocodec-ci.m4:245]: Sure, I start fixing today. I didn't know if it's needed.

Comment (singalsu) [reply tools/topology/development/sof-tgl-nocodec-ci.m4:245]: Done!

File: src/drivers/intel/dmic.c (hunk 1 above)
Comment (singalsu) [reply src/drivers/intel/dmic.c:584]: Yep, it's better.

File: src/drivers/intel/dmic.c (hunk 2 above)
Comment (singalsu) [reply src/drivers/intel/dmic.c:596]: They are not the same, with this change:

```
-       n = idx[count - 1]; /* Option with highest clock divisor and lowest mic clock rate */
+       dai_err(dai, "select_mode(): n = %d", n);
+       n = idx[found - 1]; /* Option with highest clock divisor and lowest mic clock rate */
+       dai_err(dai, "select_mode(): n = %d", n);

```

The trace output is:

```
..
[  4367496.041667] (       29.916666) c0 dmic-dai     2.0   src/drivers/intel/dmic.c:595  ERROR select_mode(): n = 0
[  4367514.583333] (       18.541666) c0 dmic-dai     2.0   src/drivers/intel/dmic.c:597  ERROR select_mode(): n = 8
..
[  4368280.250000] (       48.416668) c0 dmic-dai     2.1   src/drivers/intel/dmic.c:595  ERROR select_mode(): n = 0
[  4368299.375000] (       19.125000) c0 dmic-dai     2.1   src/drivers/intel/dmic.c:597  ERROR select_mode(): n = 3
..
```

**Comment by lgirdwood** [comment]
@singalsu is this ready for merge ? It's still a draft. I also assume it's a fix for some issues you have found ?

**Comment by singalsu** [comment]
@lgirdwood The dynamic pipelines will trigger a limitation in driver logic. 1) The possibility of dai new() to happen while another DMIC dai() is capturing wasn't assumed, and 2) the first configuration request could select configurations that make the second configuration request hard to support due to shared common resources. Issue 1) was easy to address. Issue 2) is mitigated by avoiding such as prime number decimation factors.

Seems I need to rebase this patch.

**Comment by singalsu** [comment]
Rebase is done, there were conflicts due to recent DMIC topology cleanup work. I tested this patch (though prior to rebase) yesterday with @fredoh9 and @ranj063. I found no issues in multiple pipeline capture test, and since the companion PR #3585 is also a proposal I'll change this to ready for review.

**Comment by lgirdwood** [comment]
CI showing alsabat issue that has now been fixed.

**Review by lgirdwood — APPROVED** [review]

File: tools/topology/development/sof-tgl-nocodec-ci.m4 (hunk 3 above)
Comment (lgirdwood) [reply tools/topology/development/sof-tgl-nocodec-ci.m4:247]: @ranj063 are we good now ? @plbossart changes have been merged and I dont see any conflicts.

**Comment by ranj063** [comment]
@singalsu @lgirdwood should this patch increment the ABI so that the kernel can ensure that dynamic pipes are allowed onl;y for 3.13 and beyond?

**Comment by lgirdwood** [comment]
> @singalsu @lgirdwood should this patch increment the ABI so that the kernel can ensure that dynamic pipes are allowed onl;y for 3.13 and beyond?

Done, can you link the kernel PRs here and add them to the ABI classifier project. We can then align version.

**Comment by ranj063** [comment]
corresponding Kernel PR https://github.com/thesofproject/linux/pull/2371

**Comment by ranj063** [comment]
@lgirdwood @singalsu this PR no longer needs to increment the ABI. The kernel implementation looks for the "DYNAMIC" token for pipelines to enable the feature. So this PR is good to go now!

**Review by ranj063 — APPROVED** [review]

**Comment by lgirdwood** [comment]
@singalsu can you rebase - we are good to go.

**Comment by ranj063** [comment]
@singalsu this PR is looking good now. I've tested dynamic pipelines on the up extreme with this PR and all tests passed.

**Comment by lgirdwood** [comment]
Travis - docker rate limit.

**Pull request merged**